source/io/module_javaScriptInvoke.js

```diff
@@ -155,158 +155,144 @@
-                isValidReturnType: isJavaScriptNumber
-            },
-            Double: {
-                reader: Module.eggShell.dataReadDouble,
-                writer: Module.eggShell.dataWriteDouble,
-                isValidReturnType: isJavaScriptNumber
-            },
-            String: {
-                reader: Module.eggShell.dataReadString,
-                writer: Module.eggShell.dataWriteString,
-                isValidReturnType: function (value) {
-                    return typeof value === 'string';
-                }
-            },
-            Boolean: {
-                reader: Module.eggShell.dataReadBoolean,
-                writer: Module.eggShell.dataWriteBoolean,
-                isValidReturnType: function (value) {
-                    return typeof value === 'boolean';
-                }
-            },
-            ArrayInt8: {
-                reader: Module.eggShell.dataReadTypedArray,
-                writer: Module.eggShell.dataWriteTypedArray,
-                isValidReturnType: function (value) {
-                    return value instanceof Int8Array;
-                }
-            },
-            ArrayInt16: {
-                reader: Module.eggShell.dataReadTypedArray,
-                writer: Module.eggShell.dataWriteTypedArray,
-                isValidReturnType: function (value) {
-                    return value instanceof Int16Array;
-                }
-            },
-            ArrayInt32: {
-                reader: Module.eggShell.dataReadTypedArray,
-                writer: Module.eggShell.dataWriteTypedArray,
-                isValidReturnType: function (value) {
-                    return value instanceof Int32Array;
-                }
-            },
-            ArrayUInt8: {
-                reader: Module.eggShell.dataReadTypedArray,
-                writer: Module.eggShell.dataWriteTypedArray,
-                isValidReturnType: function (value) {
-                    return value instanceof Uint8Array;
-                }
-            },
-            ArrayUInt16: {
-                reader: Module.eggShell.dataReadTypedArray,
-                writer: Module.eggShell.dataWriteTypedArray,
-                isValidReturnType: function (value) {
-                    return value instanceof Uint16Array;
-                }
-            },
-            ArrayUInt32: {
-                reader: Module.eggShell.dataReadTypedArray,
-                writer: Module.eggShell.dataWriteTypedArray,
-                isValidReturnType: function (value) {
-                    return value instanceof Uint32Array;
-                }
-            },
-            ArraySingle: {
-                reader: Module.eggShell.dataReadTypedArray,
-                writer: Module.eggShell.dataWriteTypedArray,
-                isValidReturnType: function (value) {
-                    return value instanceof Float32Array;
+        var jsInvokePeeker = function () {
+            var PeekVisitor = function () {
```

Comment:
There are some unnecessary gymnastics going on here. Go through the effort of making a prototype chain just to make a singleton is kinda weird. The constructor doesn't even do anything.

Can we just do:

```
var createPeeker = function () {
  var visitNumeric = ...

  return {
    visitBoolean = function (valueRef) {
    },
    visitInt8: visitNumeric,
    visitInt16: visitNumeric
    ...
  };
};
```

A simple module pattern

Reply:
Similar discussion for the poker

Reply:
Good point. Done.